src/WindowsTracing/KrabsTracer.cpp

```diff
@@ -43,11 +50,45 @@ void KrabsTracer::EnableProviders() {
   context_switch_provider_.add_on_event_callback(
       [this](const auto& record, const auto& context) { OnThreadEvent(record, context); });
   trace_.enable(context_switch_provider_);
+
+  stack_walk_provider_.add_on_event_callback(
+      [this](const auto& record, const auto& context) { OnStackWalkEvent(record, context); });
+  trace_.enable(stack_walk_provider_);
+}
+
+void KrabsTracer::EnableSystemProfilePrivilege(bool value) {
+  auto result = AdjustTokenPrivilege(SE_SYSTEM_PROFILE_NAME, value);
+  if (result.has_error()) ERROR("%s", result.error().message());
+}
+
+void KrabsTracer::SetupStackTracing() {
+  // Set sampling frequency for ETW trace. Note that the session handle must be 0.
+  const double frequency = capture_options_.samples_per_second();
+  CHECK(frequency > 0);
+  double period_ns = 1'000'000'000.0 / frequency;
+  static uint64_t performance_counter_period_ns = GetPerformanceCounterPeriodNs();
+  TRACE_PROFILE_INTERVAL interval = {0};
+  interval.Interval = static_cast<ULONG>(period_ns / performance_counter_period_ns);
+  ULONG status = TraceSetInformation(/*SessionHandle=*/0, TraceSampledProfileIntervalInfo,
+                                     (void*)&interval, sizeof(TRACE_PROFILE_INTERVAL));
+  CHECK(status == ERROR_SUCCESS);
+
+  // Initialize ETW stack tracing. Note that this must be executed after trace_.open() as
+  // set_trace_information needs a valid session handle.
+  CLASSIC_EVENT_ID event_id = {0};
+  event_id.EventGuid = krabs::guids::perf_info;
+  event_id.Type = kSampledProfileEventSampleProfile;
+  trace_.set_trace_information(TraceStackTracingInfo, &event_id, sizeof(CLASSIC_EVENT_ID));
+  krabs::kernel_provider stack_walk_provider(EVENT_TRACE_FLAG_PROFILE, krabs::guids::stack_walk);
+  trace_.enable(stack_walk_provider);
```

Comment:
Why is `trace_.enable(stack_walk_provider_);` called in `EnableProviders` and then `trace_.enable(stack_walk_provider);` called here with a different instance of `krabs::kernel_provider` initialized in the same way?

Reply:
Thanks for the catch, this should not be the case.